sky/packages/sky/lib/gestures/recognizer.dart

```diff
@@ -0,0 +1,141 @@
+    List<GestureArenaEntry> localEntries = new List.from(_entries);
+    _entries.clear();
+    for (GestureArenaEntry entry in localEntries)
+      entry.resolve(disposition);
+  }
+
+  void dispose() {
+    resolve(GestureDisposition.rejected);
+    for (int pointer in _trackedPointers)
+      _router.removeRoute(pointer, handleEvent);
+    _trackedPointers.clear();
+    assert(_entries.isEmpty);
+    _router = null;
+  }
+
+  void startTrackingPointer(int pointer) {
+    _router.addRoute(pointer, handleEvent);
+    _trackedPointers.add(pointer);
+    _entries.add(GestureArena.instance.add(pointer, this));
+  }
+
+  void stopTrackingPointer(int pointer) {
+    _router.removeRoute(pointer, handleEvent);
+    _trackedPointers.remove(pointer);
+    if (_trackedPointers.isEmpty)
+      didStopTrackingLastPointer();
+  }
+
+  void stopTrackingIfNotDown(sky.PointerEvent event) {
+    if (event.type == 'pointerup' || event.type == 'pointercancel')
+      stopTrackingPointer(event.pointer);
+  }
+
+}
+
+enum GestureRecognizerState {
+  ready,
+  possible,
+  defunct
+}
+
+sky.Point _getPoint(sky.PointerEvent event) {
+  return new sky.Point(event.x, event.y);
+}
+
+abstract class PrimaryPointerGestureRecognizer extends GestureRecognizer {
+  PrimaryPointerGestureRecognizer({ PointerRouter router, this.deadline })
+    : super(router: router);
+
+  final Duration deadline;
+
+  GestureRecognizerState state = GestureRecognizerState.ready;
+  int primaryPointer;
+  sky.Point initialPosition;
+  Timer _timer;
+
+  /// Override to be notified with [deadline] is exceeded.
+  void didExceedDeadline() {
+    assert(deadline == null);
+  }
+
+  /// Override to provide behavior for the primary pointer when the gesture is still possible.
+  void handlePrimaryPointer(sky.PointerEvent event);
+
+  void acceptGesture(int pointer) {
+  }
+
+  void rejectGesture(int pointer) {
+    _stopTimer();
+    if (pointer == primaryPointer)
+      state = GestureRecognizerState.defunct;
+  }
+
+  void addPointer(sky.PointerEvent event) {
+    startTrackingPointer(event.pointer);
+    if (state == GestureRecognizerState.ready) {
+      state = GestureRecognizerState.possible;
+      primaryPointer = event.pointer;
+      initialPosition = _getPoint(event);
+      if (deadline != null)
+        _timer = new Timer(deadline, didExceedDeadline);
+    }
+  }
+
+  void _stopTimer() {
+    if (_timer != null) {
+      _timer.cancel();
+      _timer = null;
+    }
+  }
+
+  void dispose() {
+    _stopTimer();
+    super.dispose();
+  }
+
+  double _getDistance(sky.PointerEvent event) {
+    sky.Offset offset = _getPoint(event) - initialPosition;
+    return offset.distance;
+  }
+
+  void handleEvent(sky.PointerEvent event) {
+    assert(state != GestureRecognizerState.ready);
+    if (state == GestureRecognizerState.possible && event.pointer == primaryPointer) {
+      if (event.type == 'pointermove' && _getDistance(event) > kTouchSlop)
```

Comment:
add a comment noting that we'll probably want to refactor the touch slop into a one-position-touch-specific subclass, since there will likely be other kinds of gesture recognisers that want to be primary-pointer-based but won't want to cancel on slop.